xdsl/transforms/canonicalization_patterns/riscv_cf.py

```diff
@@ -0,0 +1,99 @@
+        if not isinstance(op, ConditionalBranchOperation):
+            return
+
+        module: Operation | None = op.parent_op()
+        while not isinstance(module, ModuleOp) and module is not None:
+            module = module.parent_op()
+
+        if module is None:
+            return
+
+        # evaluate inputs to op
+        inputs = const_evaluate_expr_inputs(op.rs1, op.rs2)
+
+        # inputs are not compile time constant
+        if not inputs.success:
+            return
+
+        assert inputs.value is not None
+        assert len(inputs.value) == 2
+        rs1, rs2 = inputs.value
+
+        # check if the op would take the branch or not
+        branch_taken = op.const_evaluate(rs1, rs2)
+
+        # if branch is always taken, replace by jump
+        if branch_taken:
+            rewriter.replace_matched_op(
+                JOp(
+                    op.then_arguments,
+                    op.then_block,
+                    comment=f"Constant folded {op.name}",
+                )
+            )
+        # if branch is never taken, replace by "fall through"
+        else:
+            rewriter.replace_matched_op(
+                BranchOp(
+                    op.else_arguments,
+                    op.else_block,
+                    comment=f"Constant folded {op.name}",
+                )
+            )
+
+
+@dataclass
+class ConstEvalResult:
+    value: tuple[Any, ...] | None
+
+    @property
+    def success(self) -> bool:
+        return self.value is not None
+
+
+def const_evaluate_expr_inputs(*exp_inputs: SSAValue) -> ConstEvalResult:
+    """
+    Tries to evaluate the inputs to an operatin by Evaluating ConstantLike and Pure operations
+    using the interpreter.
+
+    Returns either ConstEvalResult(None), or ConstEvalResult(inputs).
+
+    If the interpretation fails, an InterpretationError is raised.
+    """
+    results: list[Any] = []
+    for operand in exp_inputs:
+        # block arguments cannot be argued about
+        if not isinstance(operand.owner, Operation):
+            return ConstEvalResult(None)
+
+        # grab the value from Li ops:
+        if isinstance(operand.owner, riscv.LiOp):
+            imm = operand.owner.immediate
+            if not isinstance(imm, IntegerAttr):
+                return ConstEvalResult(None)
+            results.append(imm.value.data)
+        # propagate through mv ops
+        elif isinstance(operand.owner, riscv.MVOp):
+            evaluated_inputs = const_evaluate_expr_inputs(*operand.owner.operands)
+            # check that the evaluation was successful
+            if not evaluated_inputs.success:
+                return ConstEvalResult(None)
+            assert evaluated_inputs.value is not None
+            results.extend(evaluated_inputs.value)
+        else:
+            # if op is neither ConstantLike nor Pure, fail
+            return ConstEvalResult(None)
+
+    return ConstEvalResult(tuple(results))
```

Comment:
It feels like this makes things more complicated than necessary. We might as well add the MVOp canonicalization pattern, and only check for LiOp directly in the rewrite, I don't think we benefit from ConstEvalResult either

Reply:
If we don't need to check for `mv` ops, then we should remove that from the filecheck. I thought following through moves was required, as it was part of the provided filecheck case.

Reply:
The filecheck is our exact use case, so we need to make it work, although we likely should add more, and better targeted ops.

The way I see to make mv work is to add a canonicalisation pattern to it, to convert it to a li. I don't see the downside of doing this, and it would improve all the other existing canonicalisation patterns.

Reply:
We have a canonicalization pattern for `mv`, but that only removes allocated moves from the the same named registers. I don't know if we can just remove mvs if they are unallocated. That feels risky and out of the scope of this PR. Let's discuss on zulip.